Add dwa support to core

kdt3rd opened the pull request:
Seems to work, although am noticing a small number of ULPs difference in the dct coding between C/C++ for some numbers, can't tell if I should be worried, or if it's just different math ops ordering causing slightly different rounding

This includes a few other (internal) refactors and fixes noticed along the way

kdt3rd committed:
Fix nodiscard warning on gcc 13

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Add buffer range checks to huf code

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Refactor cpuid code for use in dwa

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Fix issue with big endian deep encoding of chunk table

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
refactor to expose byte reordering zip compression uses for dwa

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
expose option flag to enable headers to be more easily like old style headers

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
fix potential issue with piz and not able to compress data small enough

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Initial implementation of DWAA/B compression for Core

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
reenable piz tests

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
write with legacy header to make comparing bytes in files easier

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Initial support for validating exact match in file output between C / C++

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Enable DWAA/DWAB tests

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Add new files to bazel build

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
expose c++ dwa lookup table header to core compile for bazel

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Handle o.s. c library differences

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Ensure C11 is chosen to enable alignas

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Fix function prototype mismatch

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Remove references to stdbool

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
remove further use of bool

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Revert forcing C standard

Older versions of MSVC have issues with specifying C11 for use of
stdalign.h and alignas macro. Work around this by targeting the lower
level macros / declspec

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

cary-ilm commented on src/lib/OpenEXRCore/internal_dwa_classifier.h:
```diff
@@ -0,0 +1,204 @@
+
+    if (*size <= 3) return EXR_ERR_CORRUPT_CHUNK;
+
+    //throw IEX_NAMESPACE::InputExc ("Error uncompressing DWA data"
```
dead code

kdt3rd replied:
I left them to note the error string, intending to produce a similar message via report error at some point...

cary-ilm commented on src/lib/OpenEXRCore/internal_dwa_classifier.h:
```diff
@@ -0,0 +1,204 @@
+    if (out->_cscIdx < -1 || out->_cscIdx >= 3)
+    {
+        return EXR_ERR_CORRUPT_CHUNK;
+        //throw IEX_NAMESPACE::InputExc ("Error uncompressing DWA data"
```
dead, and a few below as well

cary-ilm commented on src/lib/OpenEXRCore/internal_dwa_compressor.h:
```diff
@@ -0,0 +1,1708 @@
+    // the C++ classes used to have one buffer size for compress / uncompress
+    // but here we want to do zero-ish copy...
+    uint8_t* outBufferEnd = me->_decode->unpacked_buffer;
+    //if (outBufferSize > me->_decode->unpacked_alloc_size)
```
dead?

cary-ilm commented on src/lib/OpenEXRCore/internal_dwa_compressor.h:
```diff
@@ -0,0 +1,1708 @@
+    if (version > 2)
+    {
+        return EXR_ERR_BAD_CHUNK_LEADER;
+        //throw IEX_NAMESPACE::InputExc (
```
remove?

cary-ilm reviewed (commented):
I noted some place where you left "throw Iex" commented out from the original cpp, do you want to clean that up or is it helpful to leave in? 

Also, I may have missed some discussion, but you seem to have merged in @meshula's commits from #1385, was that intentional? That PR hasn't been committed yet, should it be closed in favor of this? 

Otherwise, I've looked this over and it looks good to me, thanks!

meshula commented:
I didn't expect that my PR would chain on to Kimball's PR like that.

I'll try closing my PR in expectation of reopening versus the main repo.

meshula commented:
Ok, closing my PR didn't do it, I guess those changes are actually committed into Kimball's tree. I have follow up work to those commits which I can open on the main repo. after this is landed. My preference would be to not land openexr-c.c and h though as they are not ready for review or distribution. I have follow up work to complete what's in that interface.

cary-ilm commented:
We'd like to include the dwa support in a 3.1.8 release, but the single-file openexr-c.c seems better suited to a 3.2/4.0 release, doesn't it?  @kdt3rd , don't you think you should back those commits out of this PR?

kdt3rd commented:
@cary-ilm yes, I did not actually accept Nick's PR on my branch, it's interesting that it is chaining up to here... ah, the vagaries of github PRs...

meshula commented:
Well it's even weirder because my PR was on the main repo, not yours! I have a different PR on your repo, but that has commits not appearing here. I've closed both of them in hopes of untangling things.

kdt3rd commented:
fyi, have confirmed Nick's commits are not in my local branch at all. I need to rebase onto a branch that removes the old mac os build that keeps timing out, so that should process should re-clean the history. I have one pending fix to push in the dwa support, and another difference in the decoder side to track down, then I think this will be ready to go

kdt3rd committed:
Add unit test for x86 cpu id checks

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Fix implementation of cpuid

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Fix constructor typo

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Add alignment helper utility

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Restore exactness checks for DWAA/DWAB

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
minor optimization

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
clarify quantization step variables

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
clean up inline asm to avoid issues with clang format

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Add missing include guards

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Fix pedantic warnings

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Expose config header to core build

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Cleanup dead code, fix compile warnings

Core should now compile with
-Weverything
-Wno-reserved-identifier
-Wno-covered-switch-default
-Wno-cast-align
-Wno-overlength-strings

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Fix warnings on other platforms than linux

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

kdt3rd committed:
Fix straggling warnings on MacOS

Signed-off-by: Kimball Thurston <kdt3rd@gmail.com>

meshula commented:
Looks like you hit all the structural points I bumped into when I put openexr-c.c together. I'll start work to bounce it off our internal build system on the assumption that this is ready to land.

meshula reviewed (approved):
This looks good ~ I've already run your previous iteration through our internal build system, and got it working loading textures and sent 'em to gpu. I had a pile of "pedantic" and other clean ups, seems like you've hit the lot, with luck I can strip my clean ups in favor of using these commits "raw". I think it's ready to land, on the basis of the passing unit tests, and on the basis of my local exr reader working properly with your previous iteration.
